File: packages/react-native-enriched-markdown/ios/renderer/ListItemRenderer.m
#import "ListItemRenderer.h"
#import "ENRMUIKit.h"
#import "LastElementUtils.h"
#import "MarkdownASTNode.h"
#import "ParagraphStyleUtils.h"
#import "RenderContext.h"
#import "RendererFactory.h"
#import "StyleConfig.h"

NSString *const ListDepthAttribute = @"ListDepth";
NSString *const ListTypeAttribute = @"ListType";
NSString *const ListItemNumberAttribute = @"ListItemNumber";
NSString *const ListItemMarkerStartAttribute = @"ListItemMarkerStart";
NSString *const TaskItemAttribute = @"TaskItem";
NSString *const TaskCheckedAttribute = @"TaskChecked";
NSString *const TaskIndexAttribute = @"TaskIndex";

@interface ListItemRenderer ()
- (void)applyCheckedDecorationsTo:(NSMutableAttributedString *)output
                            range:(NSRange)range
                     nestingLevel:(NSInteger)nestingLevel;
@end

@implementation ListItemRenderer

- (void)renderNodeContent:(MarkdownASTNode *)node
                     into:(NSMutableAttributedString *)output
                  context:(RenderContext *)context
{
  if (!context)
    return;

  context.listItemNumber++;
  const NSInteger currentPosition = context.listItemNumber;
  const NSInteger currentDepth = context.listDepth; // 1-based (1 = top level)

  const BOOL isTask = [node.attributes[@"isTask"] isEqualToString:@"true"];
  const BOOL isChecked = isTask && [node.attributes[@"taskChecked"] isEqualToString:@"true"];
  NSInteger taskIndex = -1;
  if (isTask) {
    taskIndex = context.taskItemCount;
    context.taskItemCount++;
  }

  const NSUInteger startLocation = output.length;

  // Render the actual content of the list item (text, bolding, etc.)
  [_rendererFactory renderChildrenOfNode:node into:output context:context];

  // Ensure every list item ends with a newline to prevent paragraph merging
  if (output.length > startLocation && ![output.string hasSuffix:@"\n"]) {
    [output appendAttributedString:kNewlineAttributedString];
  }

  const NSRange itemRange = NSMakeRange(startLocation, output.length - startLocation);
  if (itemRange.length == 0)
    return;

  // Informs MarkdownAccessibilityElementBuilder about the specific boundaries of this list item
  [context registerListItemRange:itemRange
                        position:currentPosition
                           depth:currentDepth
                       isOrdered:(context.listType == ListTypeOrdered)];

  // currentDepth - 1 handles the horizontal offset for nested lists
  const NSInteger nestingLevel = currentDepth - 1;
  const CGFloat baseMarkerWidth = isTask                                  ? [_config effectiveListMarginLeftForTask]
                                  : (context.listType == ListTypeOrdered) ? [_config effectiveListMarginLeftForNumber]
                                                                          : [_config effectiveListMarginLeftForBullet];

  const CGFloat totalIndent =
      baseMarkerWidth + [_config effectiveListGapWidth] + (nestingLevel * [_config listStyleMarginLeft]);

  const CGFloat lineHeightConfig = [_config listStyleLineHeight];

  // Boxing metadata for attributed string storage
  NSMutableDictionary *metadata = [@{
    ListDepthAttribute : @(nestingLevel),
    ListTypeAttribute : @(context.listType),
    ListItemNumberAttribute : @(currentPosition)
  } mutableCopy];

  if (isTask) {
    metadata[TaskItemAttribute] = @YES;
    metadata[TaskCheckedAttribute] = @(isChecked);
    metadata[TaskIndexAttribute] = @(taskIndex);
  }

  // Preserve styles of nested sub-lists and code blocks by applying the list
  // paragraph style only to the surviving gaps between them.
  NSMutableArray<NSValue *> *skipRanges = [NSMutableArray array];
  [output enumerateAttribute:CodeBlockAttributeName
                     inRange:itemRange
                     options:0
                  usingBlock:^(id value, NSRange range, BOOL *stop) {
                    if ([value boolValue]) {
                      [skipRanges addObject:[NSValue valueWithRange:range]];
                    }
                  }];
  [output enumerateAttribute:ListDepthAttribute
                     inRange:itemRange
                     options:0
                  usingBlock:^(id depthAttr, NSRange range, BOOL *stop) {
                    if (depthAttr && [depthAttr integerValue] > nestingLevel) {
                      [skipRanges addObject:[NSValue valueWithRange:range]];
                    }
                  }];
  [skipRanges sortUsingComparator:^NSComparisonResult(NSValue *a, NSValue *b) {
    NSUInteger la = [a rangeValue].location;
    NSUInteger lb = [b rangeValue].location;
    return la < lb ? NSOrderedAscending : (la > lb ? NSOrderedDescending : NSOrderedSame);
  }];

  void (^applyStyleToRange)(NSRange) = ^(NSRange range) {
    if (range.length == 0)
      return;
    NSMutableParagraphStyle *style = [[NSMutableParagraphStyle alloc] init];
    style.firstLineHeadIndent = totalIndent;
    style.headIndent = totalIndent;
    if (lineHeightConfig > 0) {
      style.minimumLineHeight = lineHeightConfig;
      style.maximumLineHeight = lineHeightConfig;
    }
    NSMutableDictionary *attributesToApply = [metadata mutableCopy];
    attributesToApply[NSParagraphStyleAttributeName] = style;
    [output addAttributes:attributesToApply range:range];
  };

  NSUInteger pos = itemRange.location;
  const NSUInteger itemEnd = NSMaxRange(itemRange);
  for (NSValue *val in skipRanges) {
    NSRange skip = [val rangeValue];
    if (pos < skip.location) {
      applyStyleToRange(NSMakeRange(pos, skip.location - pos));
    }
    pos = MAX(pos, NSMaxRange(skip));
  }
  if (pos < itemEnd) {
    applyStyleToRange(NSMakeRange(pos, itemEnd - pos));
  }

  [output addAttribute:ListItemMarkerStartAttribute value:@YES range:NSMakeRange(startLocation, 1)];

  if (isTask && isChecked) {
    [self applyCheckedDecorationsTo:output range:itemRange nestingLevel:nestingLevel];
  }
}

- (void)applyCheckedDecorationsTo:(NSMutableAttributedString *)output
                            range:(NSRange)range
                     nestingLevel:(NSInteger)nestingLevel
{
  RCTUIColor *checkedColor = [_config taskListCheckedTextColor];
  BOOL shouldStrikethrough = [_config taskListCheckedStrikethrough];

  if (!checkedColor && !shouldStrikethrough) {
    return;
  }

  NSMutableDictionary *checkedAttrs = [NSMutableDictionary dictionary];

  if (checkedColor) {
    checkedAttrs[NSForegroundColorAttributeName] = checkedColor;
  }

  if (shouldStrikethrough) {
    checkedAttrs[NSStrikethroughStyleAttributeName] = @(NSUnderlineStyleSingle);

    RCTUIColor *lineColor = checkedColor ?: [_config listStyleColor];
    if (lineColor) {
      checkedAttrs[NSStrikethroughColorAttributeName] = lineColor;
    }
  }

  [output enumerateAttribute:ListDepthAttribute
                     inRange:range
                     options:0
                  usingBlock:^(NSNumber *depth, NSRange segmentRange, BOOL *stop) {
                    BOOL isNestedSubItem = (depth && [depth integerValue] > nestingLevel);
                    if (isNestedSubItem) {
                      return;
                    }

                    // Skip code block ranges — preserve CodeBlockRenderer styles.
                    NSNumber *isCodeBlock = [output attribute:CodeBlockAttributeName
                                                      atIndex:segmentRange.location
                                               effectiveRange:nil];
                    if ([isCodeBlock boolValue]) {
                      return;
                    }

                    [output addAttributes:checkedAttrs range:segmentRange];
                  }];
}

@end

File: packages/ios-enriched-markdown/Sources/EnrichedMarkdown/Rendering/Renderers/ListItemRenderer.swift
import UIKit

final class ListItemRenderer: NodeRenderer {
    private let factory: RendererFactory
    private let config: MarkdownStyleConfig

    init(factory: RendererFactory, config: MarkdownStyleConfig) {
        self.factory = factory
        self.config = config
    }

    func render(node: MarkdownASTNode, into output: NSMutableAttributedString, context: RenderContext) {
        context.listItemNumber += 1
        let currentPosition = context.listItemNumber
        let currentDepth = context.listDepth
        let nestingLevel = currentDepth - 1

        let startLocation = output.length
        factory.renderChildren(of: node, into: output, context: context)
        ParagraphStyleHelpers.ensureTrailingNewline(in: output)

        let itemRange = NSRange(location: startLocation, length: output.length - startLocation)
        guard itemRange.length > 0 else { return }

        let baseMarkerWidth = effectiveMarkerWidth(for: context.listType)
        let gapWidth = max(config.list.gapWidth ?? 12, 4)
        let marginLeft = config.list.marginLeft ?? 24
        let totalIndent = baseMarkerWidth + gapWidth + (CGFloat(nestingLevel) * marginLeft)
        let lineHeight = config.list.lineHeight ?? 0

        let metadata: [NSAttributedString.Key: Any] = [
            MarkdownAttribute.listDepth: nestingLevel,
            MarkdownAttribute.listType: context.listType.rawValue,
            MarkdownAttribute.listItemNumber: currentPosition
        ]

        applyListItemStyling(
            to: output,
            itemRange: itemRange,
            nestingLevel: nestingLevel,
            metadata: metadata,
            totalIndent: totalIndent,
            lineHeight: lineHeight
        )
    }

    private func effectiveMarkerWidth(for listType: ListType) -> CGFloat {
        let minWidth = config.list.markerMinWidth ?? 0
        switch listType {
        case .ordered:
            return max(minWidth, 20)
        case .unordered:
            return max(minWidth, config.list.bulletSize ?? 6)
        }
    }

    private func applyListItemStyling(
        to output: NSMutableAttributedString,
        itemRange: NSRange,
        nestingLevel: Int,
        metadata: [NSAttributedString.Key: Any],
        totalIndent: CGFloat,
        lineHeight: CGFloat
    ) {
        let string = output.string as NSString
        var location = itemRange.location
        let end = NSMaxRange(itemRange)

        while location < end {
            let paragraphRange = string.paragraphRange(for: NSRange(location: location, length: 0))
            let applyRange = NSIntersectionRange(paragraphRange, itemRange)
            guard applyRange.length > 0 else { break }

            if shouldSkipListStyling(in: output, range: applyRange, nestingLevel: nestingLevel) {
                location = NSMaxRange(applyRange)
                continue
            }

            let style = NSMutableParagraphStyle()
            style.firstLineHeadIndent = totalIndent
            style.headIndent = totalIndent

            if lineHeight > 0 {
                style.minimumLineHeight = lineHeight
                style.maximumLineHeight = lineHeight
            }

            var attributes = metadata
            attributes[.paragraphStyle] = style
            output.addAttributes(attributes, range: applyRange)

            if lineHeight > 0 {
                ParagraphStyleHelpers.applyBaselineOffset(to: output, range: applyRange)
            }

            location = NSMaxRange(applyRange)
        }
    }

    private func shouldSkipListStyling(
        in output: NSMutableAttributedString,
        range: NSRange,
        nestingLevel: Int
    ) -> Bool {
        if let depth = MarkdownAttributeValue.intValue(
            from: output.attribute(MarkdownAttribute.listDepth, at: range.location, effectiveRange: nil)
        ), depth > nestingLevel {
            return true
        }

        if MarkdownAttributeValue.boolValue(
            from: output.attribute(MarkdownAttribute.codeBlock, at: range.location, effectiveRange: nil)
        ) {
            return true
        }

        return false
    }
}
